### PR TITLE
IDE0004 Redundant cast: False positive for IntPtr -> nint -> int

### LINKED ISSUE #51123: Convert IntPtr to nint to int incorrectly reported as redundant
**Version Used**: 
Microsoft Visual Studio Community 2019
Version 16.8.4
**Steps to Reproduce**:
```C#
var x = (IntPtr)long.MaxValue; // Only works in 64bit platform. OverflowException will be thrown in 32bit platform
//var y = (int)x; // OverflowException
var y = (int)(nint)x; // IDE0004 says (nint) is redundant
```
**Expected Behavior**:
`(nint)` is not reported as redundant
**Actual Behavior**:
`(nint)` is reported as redundant

### PR DESCRIPTION
**Pull request by MaStr11**
Fixes #51123

### CODE REVIEW

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Simplification/Simplifiers/CastSimplifier.cs [hunk 1]
```diff
@@ -528,14 +528,22 @@ private static bool IsIntPtrToNativeIntegerNestedCast(ExpressionSyntax castNode,
...
                 {
-                    return true;
+                    // Do we have a parent cast to the opposite?
+                    if (parentCastType.IsNativeIntegerType &&
+                        parentCastType.SpecialType == oppositeType)
+                    {
+                        return true;
+                    }
+                    // cases like (int)(nint)myIntPtr
+                    if (parentCastType.IsNumericType())
```
Comment (MaStr11) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Simplification/Simplifiers/CastSimplifier.cs:543]: This is probably too broad. `IsNumeric` applies to
```CS
                switch (type.SpecialType)
                {
                    case SpecialType.System_Byte:
                    case SpecialType.System_SByte:
                    case SpecialType.System_Int16:
                    case SpecialType.System_UInt16:
                    case SpecialType.System_Int32:
                    case SpecialType.System_UInt32:
                    case SpecialType.System_Int64:
                    case SpecialType.System_UInt64:
                    case SpecialType.System_Single:
                    case SpecialType.System_Double:
                    case SpecialType.System_Decimal:
                    case SpecialType.System_IntPtr when type.IsNativeIntegerType:
                    case SpecialType.System_UIntPtr when type.IsNativeIntegerType:
                        return true;
                }
```
It also makes the previous check `parentCastType.IsNativeIntegerType && parentCastType.SpecialType == oppositeType` redundant (oppositeType is either IntPtr or UIntPtr).

File: src/Analyzers/CSharp/Tests/RemoveUnnecessaryCast/RemoveUnnecessaryCastTests.cs [hunk 2]
```diff
@@ -8207,6 +8207,27 @@ public class C {
             await test.RunAsync();
         }
 
+        [WorkItem(51123, "https://github.com/dotnet/roslyn/issues/51123")]
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsRemoveUnnecessaryCast)]
+        public async Task DoRemoveNativeIntCastsToInt()
```
Comment (Youssef1313) [inline src/Analyzers/CSharp/Tests/RemoveUnnecessaryCast/RemoveUnnecessaryCastTests.cs:8212]: Did you mean "DoNot"?

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Simplification/Simplifiers/CastSimplifier.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Simplification/Simplifiers/CastSimplifier.cs:543]: Being too broad is ok here. Better to leave in some unnecessary casts than remove some necessary ones.

**Review by CyrusNajmabadi — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
Thanks!

**Pull request merged**